### PR TITLE
Feat/collectionbadges

### LINKED ISSUE #14: Badges: Collection badges Model
Define Model to store collection badges

### PR DESCRIPTION
**Pull request by jona159**
- add models, serializers and views for collectionbadges and superbadges
- register them in admin panel
- create public endpoints to fetch all collectionbadges and superbagdes
- create endpoints for authenticated users to fetch single collectionbadges and superbadges

closes #14

### CODE REVIEW

File: apps/issuer/admin.py [hunk 1]
```diff
@@ -8,10 +8,86 @@
...
+
+    def save_model(self, request, obj, form, change):
+        force_resize = False
+        if 'image' in form.changed_data:
+            force_resize = True
+        obj.save(force_resize=force_resize)
+
+    def badge_image(self, obj):
+        return mark_safe('<img src="{}" width="32"/>'.format(obj.image.url)) if obj.image else ''
+    badge_image.short_description = 'Badge'
```
Comment (timber-they) [inline apps/issuer/admin.py:47]: Is this description so helpful?

File: apps/issuer/admin.py [hunk 2]
```diff
@@ -8,10 +8,86 @@
...
+    
+    pass
+
+
+
+class SuperBadgeInstanceInline(TabularInline):
+    model = SuperBadge.assertions.through
+    extra = 0
+    # raw_id_fields = ('badgeinstance',)
```
Comment (timber-they) [inline apps/issuer/admin.py:56]: Commented code

File: apps/issuer/admin.py [hunk 3]
```diff
@@ -8,10 +8,86 @@
...
+
+class SuperBadgeAdmin(ModelAdmin):
```
Comment (timber-they) [inline apps/issuer/admin.py:58]: Isn't this exactly the same as `CollectionBadgeAdmin`? Maybe the duplicate code can be extracted into a super class?

File: apps/issuer/api.py [hunk 4]
```diff
@@ -99,6 +99,59 @@ def put(self, request, **kwargs):
     def delete(self, request, **kwargs):
         return super(IssuerDetail, self).delete(request, **kwargs)
 
+class CollectionBadgeDetail(BaseEntityDetailView):
+    model = CollectionBadgeContainer
+    v1_serializer_class = CollectionBadgeClassSerializerV1
+    # v2_serializer_class = CollectionBadgeClassSerializerV2
+    permission_classes = (AuthenticatedWithVerifiedIdentifier, AuditedModelOwner)
+    valid_scopes = {
+        'get': ['r:issuer', 'rw:issuer'],
```
Comment (timber-they) [inline apps/issuer/api.py:108]: Why do the scopes refer to issuers?

File: apps/issuer/api.py [hunk 5]
```diff
@@ -99,6 +99,59 @@ def put(self, request, **kwargs):
...
+        'post': ['rw:issuer'],
+        'put': ['rw:issuer'],
+        'delete': ['rw:issuer']
+    }
+
+    @apispec_get_operation('CollectionBadge',
+                           summary='Get a single CollectionBadge',
+                           tags=['Issuer']
```
Comment (timber-they) [inline apps/issuer/api.py:116]: Why is the tag `Issuer` and not something referring to collection badges?
This also applies to all the cases below.

File: apps/issuer/api.py [hunk 6]
```diff
@@ -99,6 +99,59 @@ def put(self, request, **kwargs):
...
+
+
+
+class AllSuperBadgeClassesList(UncachedPaginatedViewMixin, BaseEntityListView):
+    """
+    GET a list of superbadgeclasses 
+    """
+    model = SuperBadge
+    v1_serializer_class = SuperBadgeClassSerializerV1
+    v2_serializer_class = SuperBadgeClassSerializerV2
```
Comment (timber-they) [inline apps/issuer/api.py:143]: Is it on purpose that here (in contrast to above) the `v2` isn't commented out?

File: apps/issuer/api.py [hunk 7]
```diff
@@ -99,6 +99,59 @@ def put(self, request, **kwargs):
...
+    valid_scopes = ["rw:serverAdmin"]
```
Comment (timber-they) [inline apps/issuer/api.py:144]: How does this scope make sense?

File: apps/issuer/api.py [hunk 8]
```diff
@@ -143,6 +196,41 @@ def post(self, request, **kwargs):
...
+    """
+    model = CollectionBadgeContainer
+    permission_classes = [
+        IsServerAdmin
+        | (AuthenticatedWithVerifiedIdentifier & IsEditor & BadgrOAuthTokenHasScope)
+        | BadgrOAuthTokenHasEntityScope
+    ]
+    v1_serializer_class = CollectionBadgeClassSerializerV1
+    v2_serializer_class = CollectionBadgeClassSerializerV2
+    valid_scopes = ["rw:issuer"]
```
Comment (timber-they) [inline apps/issuer/api.py:212]: Same as above

File: apps/issuer/api.py [hunk 9]
```diff
@@ -143,6 +196,41 @@ def post(self, request, **kwargs):
...
 
 
+class AllCollectionBadgeClassesList(UncachedPaginatedViewMixin, BaseEntityListView):
+    """
+    GET a list of collectionbadgeclasses within one issuer context or
+    POST to create a new collectionbadgeclass within the issuer context
+    """
+    model = CollectionBadgeContainer
+    permission_classes = [
+        IsServerAdmin
```
Comment (timber-they) [inline apps/issuer/api.py:206]: Why does one have to be server admin?

File: apps/issuer/api.py [hunk 10]
```diff
@@ -143,6 +196,41 @@ def post(self, request, **kwargs):
...
+        summary="Get a list of CollectionBadgeClasses for authenticated user",
+        tags=["CollectionBadgeClasses"],
+    )
+    def get(self, request, **kwargs):
+        return super(AllCollectionBadgeClassesList, self).get(request, **kwargs)
+
+    @apispec_post_operation('CollectionBadgeClass',
+        summary="Create a new CollectionBadgeClass",
+        tags=["CollectionBadgeClasses"],
+        # TODO: Add parameters
```
Comment (timber-they) [inline apps/issuer/api.py:227]: Is this TODO still up to date?

File: apps/issuer/api.py [hunk 11]
```diff
@@ -246,6 +334,47 @@ def delete(self, request, **kwargs):
...
+    """
+    model = SuperBadge
+    permission_classes = [
+        IsServerAdmin
+        | (AuthenticatedWithVerifiedIdentifier )
+    ]
+    v1_serializer_class = SuperBadgeClassSerializerV1
+    v2_serializer_class = SuperBadgeClassSerializerV2
+
+    valid_scopes = ["rw:all", "rw:issuer:*"]
```
Comment (timber-they) [inline apps/issuer/api.py:349]: Same as above

File: apps/issuer/api.py [hunk 12]
```diff
@@ -246,6 +334,47 @@ def delete(self, request, **kwargs):
...
+    """
+    model = CollectionBadgeContainer
+    permission_classes = [
+        IsServerAdmin
+        | (AuthenticatedWithVerifiedIdentifier)
+    ]
+    v1_serializer_class = CollectionBadgeClassSerializerV1
+    v2_serializer_class = CollectionBadgeClassSerializerV2
+
+    valid_scopes = ["rw:all", "rw:issuer:*"]
```
Comment (timber-they) [inline apps/issuer/api.py:370]: Same as above

**Review by timber-they — CHANGES_REQUESTED** [review]
For now I only managed to review `apps/issuer/admin.py` and `apps/issuer/api.py`. I'm gonna continue with the rest later, but you can already check out the changes I requested.

File: apps/issuer/api.py (hunk 12 above)
Comment (jona159) [reply apps/issuer/api.py]: Update ``` rw:all``` to ```rw:issuer```

File: apps/issuer/api.py (hunk 11 above)
Comment (jona159) [reply apps/issuer/api.py]: Update ``` rw:all``` to ```rw:issuer```

File: apps/issuer/api.py (hunk 10 above)
Comment (jona159) [reply apps/issuer/api.py:227]: yes

File: apps/issuer/api.py (hunk 9 above)
Comment (jona159) [reply apps/issuer/api.py:206]: You`re right, i extended the permission classes

File: apps/issuer/api.py (hunk 5 above)
Comment (jona159) [reply apps/issuer/api.py]: Updated the tags to ``` CollectionBadges```

File: apps/issuer/api.py (hunk 4 above)
Comment (jona159) [reply apps/issuer/api.py]: I am not sure tbh, but the badgeclasses have the same scope and i tried to stay consistent with previously written code. Do you have another suggestion for the scopes?

File: apps/issuer/admin.py (hunk 3 above)
Comment (jona159) [reply apps/issuer/admin.py:57]: I dont necessarily expect the classes to stay the same. I think at the moment it makes sense to keep them seperate, but if they dont diverge later on i will create a super class.

File: apps/issuer/admin.py (hunk 2 above)
Comment (jona159) [reply apps/issuer/admin.py]: I removed the comment

File: apps/issuer/admin.py (hunk 1 above)
Comment (jona159) [reply apps/issuer/admin.py]: Updated it to ```CollectionBadge```, dont know what else to put here at the moment

File: apps/issuer/api.py (hunk 6 above)
Comment (jona159) [reply apps/issuer/api.py:143]: Commenting the ```CollectionBadgeClassSerializerV2``` above was not on purpose, i uncommented it.

File: apps/issuer/api.py (hunk 7 above)
Comment (jona159) [reply apps/issuer/api.py]: Updated it to ```rw:issuer```

File: apps/issuer/api.py (hunk 4 above)
Comment (timber-they) [reply apps/issuer/api.py]: Hm weird, maybe at some point we need to give the scopes a general rewrite, but then it's probably fine for now

File: apps/issuer/api.py (hunk 10 above)
Comment (timber-they) [reply apps/issuer/api.py:227]: And should it be done before being merged?

File: apps/issuer/serializers_v1.py [hunk 13]
```diff
@@ -223,6 +225,122 @@ class BadgeClassExpirationSerializerV1(serializers.Serializer):
     duration = serializers.ChoiceField(source='expires_duration', allow_null=True,
                                        choices=BadgeClass.EXPIRES_DURATION_CHOICES)
 
+# https://stackoverflow.com/questions/50973569/django-rest-framework-relatedfield-cant-return-a-dict-object
+
+class ModifiedRelatedField(serializers.RelatedField):
+    def get_choices(self, cutoff=None):
```
Comment (timber-they) [inline apps/issuer/serializers_v1.py:231]: To me it's not intuitive what exactly this method is doing (ordering something?) and what it's returning (sometimes an empty object, sometimes an ordered dictionary?). Maybe both a comment documenting this and a return type definition would be nice. And maybe it makes sense to not return an empty object, but rather an empty `OrderedDict`?

File: apps/issuer/serializers_v1.py [hunk 14]
```diff
@@ -223,6 +225,122 @@ class BadgeClassExpirationSerializerV1(serializers.Serializer):
...
+
+    badges = SuperBadgeBadgeClassField(
+        queryset=BadgeClass.objects.all(), many=True, source='cached_badgeclasses'
+    )
+
+    def to_representation(self, instance):
+        representation = super(SuperBadgeClassSerializerV1, self).to_representation(instance)
+        return representation
+
+    def validate_image(self, image):
```
Comment (timber-they) [inline apps/issuer/serializers_v1.py:273]: This clearly doesn't just validate the image, but change the name and returns the image

File: apps/issuer/serializers_v1.py [hunk 15]
```diff
@@ -223,6 +225,122 @@ class BadgeClassExpirationSerializerV1(serializers.Serializer):
...
+            allow_null=True, max_length=255)
+    image = ValidImageField(required=False)
+    badges = CollectionBadgeBadgeClassField(many=True, queryset=BadgeClass.objects.all(), source='cached_collects')
+    slug = StripTagsCharField(required=False, max_length=128, source='entity_id')
+
+    def to_representation(self, instance):
+        representation = super(CollectionBadgeClassSerializerV1, self).to_representation(instance)
+        return representation
+    
+    def validate_image(self, image):
```
Comment (timber-they) [inline apps/issuer/serializers_v1.py:322]: Same as above

File: apps/issuer/serializers_v2.py [hunk 16]
```diff
@@ -75,7 +75,7 @@ class IssuerSerializerV2(DetailSerializerV2, OriginalJsonSerializerMixin):
     createdBy = EntityRelatedFieldV2(source='cached_creator', queryset=BadgeUser.cached, required=False)
     name = StripTagsCharField(max_length=1024)
     image = ValidImageField(required=False, use_public=True, source='*')
-    email = serializers.EmailField(max_length=255, required=True)
+    email = serializers.EmailField(max_length=255, required=True)
```
Comment (timber-they) [inline apps/issuer/serializers_v2.py:78]: Why is there an extra space here?

File: apps/issuer/serializers_v2.py [hunk 17]
```diff
@@ -411,6 +411,191 @@ def create(self, validated_data):
 
         return super(BadgeClassSerializerV2, self).create(validated_data)
 
+class ModifiedRelatedField(serializers.RelatedField):
```
Comment (timber-they) [inline apps/issuer/serializers_v2.py:414]: Same as above

File: apps/issuer/serializers_v2.py [hunk 18]
```diff
@@ -411,6 +411,191 @@ def create(self, validated_data):
...
+                }),
+                ('image', {
+                    'type': "string",
+                    'format': "data:image/png;base64",
+                    'description': "Base64 encoded string of an image that represents the CollectionBadgeClass.",
+                    'required': False,
+                }),
+                ('assertions', {
+                    'type': "array",
+                    # 'items': {
```
Comment (timber-they) [inline apps/issuer/serializers_v2.py:489]: What is this comment doing here?

File: apps/issuer/serializers_v2.py [hunk 19]
```diff
@@ -411,6 +411,191 @@ def create(self, validated_data):
...
+                    'description': "Unique identifier for this SuperBadgeClass",
+                    'readOnly': True,
+                }),
+                ('entityType', {
+                    'type': "string",
+                    'format': "string",
+                    'description': "\"SuperBadgeClass\"",
+                    'readOnly': True,
+                }),
+                # ('createdAt', {
```
Comment (timber-they) [inline apps/issuer/serializers_v2.py:542]: What is this comment doing here?

File: apps/issuer/serializers_v2.py [hunk 20]
```diff
@@ -411,6 +411,191 @@ def create(self, validated_data):
...
+                }),
+                ('image', {
+                    'type': "string",
+                    'format': "data:image/png;base64",
+                    'description': "Base64 encoded string of an image that represents the SuperBadgeClass.",
+                    'required': False,
+                }),
+                ('assertions', {
+                    'type': "array",
+                    # 'items': {
```
Comment (timber-they) [inline apps/issuer/serializers_v2.py:582]: Again

File: apps/issuer/v1_api_urls.py [hunk 21]
```diff
@@ -10,6 +10,15 @@
     url(r'^all-badges$', AllBadgeClassesList.as_view(), name='v1_api_issuer_all_badges_list'),
     url(r'^all-badges/find$', FindBadgeClassDetail.as_view(), name='v1_api_find_badgeclass_by_identifier'),
 
+    url(r'^all-superbadges$', AllSuperBadgeClassesList.as_view(), name='v1_api_issuer_all_superbadges_list'),
+    url(r'^all-collectionbadges$', AllCollectionBadgeClassesList.as_view(), name='v1_api_issuer_all_collectionbadges_list'),
+
+    url(r'^collectionbadges/(?P<slug>[-\w]+)$', CollectionBadgeDetail.as_view(),
+        name='v1_api_collectionbadges_detail'),
+
+    # url(r'^superbadges/(?P<slug>[-\w]+)$', SuperBadgeDetail.as_view(),
```
Comment (timber-they) [inline apps/issuer/v1_api_urls.py:19]: Comment?

**Review by timber-they — CHANGES_REQUESTED** [review]

File: apps/issuer/v1_api_urls.py (hunk 21 above)
Comment (jona159) [reply apps/issuer/v1_api_urls.py:20]: This endpoint is not ready for use yet, the ```SuperBadgeDetail``` class is not implemented yet. I added a TODO comment to clarify this

File: apps/issuer/serializers_v2.py (hunk 19 above)
Comment (jona159) [reply apps/issuer/serializers_v2.py]: Removed

File: apps/issuer/serializers_v2.py (hunk 18 above)
Comment (jona159) [reply apps/issuer/serializers_v2.py]: Removed

File: apps/issuer/serializers_v2.py (hunk 16 above)
Comment (jona159) [reply apps/issuer/serializers_v2.py]: Idk, i remove it now

File: apps/issuer/serializers_v1.py (hunk 14 above)
Comment (jona159) [reply apps/issuer/serializers_v1.py:279]: I copied the function from another place in the code. I would agree though that renaming the function would make sense, but i think then we should do it in all places

File: apps/issuer/serializers_v1.py [hunk 22]
```diff
@@ -223,6 +225,122 @@ class BadgeClassExpirationSerializerV1(serializers.Serializer):
     duration = serializers.ChoiceField(source='expires_duration', allow_null=True,
                                        choices=BadgeClass.EXPIRES_DURATION_CHOICES)
 
+# https://stackoverflow.com/questions/50973569/django-rest-framework-relatedfield-cant-return-a-dict-object
```
[the remark this answers is not in the document]
Comment (jona159) [reply apps/issuer/serializers_v1.py]: I tried documenting what the class and the method does via some comments now.

File: apps/issuer/models.py [hunk 23]
```diff
@@ -564,7 +564,105 @@ def get_user_or_none(recipient_id, recipient_type):
...
+                               BaseVersionedEntity):
+    entity_class_name = 'CollectionBadgeContainer'
+    name = models.CharField(max_length=128)
+    description = models.CharField(max_length=255, blank=True)
 
+    image = models.FileField(upload_to='uploads/badges', blank=True)
+    image_preview = models.FileField(upload_to='uploads/badges', blank=True, null=True)
+
+    
+    # slug = models.CharField(max_length=254, blank=True, null=True, default=None)
```
[the remark this answers is not in the document]
Comment (jona159) [reply apps/issuer/models.py]: Removed the comment

File: apps/issuer/migrations/0065_auto_20240220_1218.py [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('badgeclass', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='issuer.BadgeClass')),
+                ('superbadge', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='issuer.SuperBadge')),
+            ],
+            options={
+                'abstract': False,
+            },
+        ),
+        migrations.AddField(
+            model_name='superbadge',
```
[the remark this answers is not in the document]
Comment (jona159) [reply apps/issuer/migrations/0065_auto_20240220_1218.py:43]: I did not write the lowercase version here, all the code in the migrations folder is generated automatically when running migrations. Or did you mean something else?

File: apps/issuer/api.py (hunk 10 above)
Comment (jona159) [reply apps/issuer/api.py]: Actually, it's not really needed for now i guess. I removed it for now

File: apps/issuer/migrations/0065_auto_20240220_1218.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (timber-they) [reply apps/issuer/migrations/0065_auto_20240220_1218.py:43]: Hm ok, I guess it's fine then

File: apps/issuer/serializers_v1.py (hunk 14 above)
Comment (timber-they) [reply apps/issuer/serializers_v1.py:279]: Do you think it would make sense to do this in the scope of this PR?

**Review by timber-they — APPROVED** [review]
Then I think you can merge it

**Review by timber-they — APPROVED** [review]

**Pull request merged**